x/rollapp/keeper/fraud_handler.go

```diff
@@ -0,0 +1,121 @@
+package keeper
+
+import (
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+
+	clienttypes "github.com/cosmos/ibc-go/v6/modules/core/02-client/types"
+	tmtypes "github.com/cosmos/ibc-go/v6/modules/light-clients/07-tendermint/types"
+
+	common "github.com/dymensionxyz/dymension/v3/x/common/types"
+	"github.com/dymensionxyz/dymension/v3/x/rollapp/types"
+)
+
+// HandleFraud handles the fraud evidence submitted by the user.
+func (k Keeper) HandleFraud(ctx sdk.Context, rollappID, clientId string, fraudHeight uint64, seqAddr string) error {
+	// Get the rollapp from the store
+	rollapp, found := k.GetRollapp(ctx, rollappID)
+	if !found {
+		return sdkerrors.Wrapf(types.ErrInvalidRollappID, "rollapp with ID %s not found", rollappID)
+	}
+
+	stateInfo, err := k.FindStateInfoByHeight(ctx, rollappID, fraudHeight)
+	if err != nil {
+		return err
+	}
+
+	//check height is not finalized
+	if stateInfo.Status == common.Status_FINALIZED {
+		return sdkerrors.Wrapf(types.ErrDisputeAlreadyFinalized, "state info for height %d is already finalized", fraudHeight)
+	}
+
+	//check height is not reverted
+	if stateInfo.Status == common.Status_REVERTED {
+		return sdkerrors.Wrapf(types.ErrDisputeAlreadyReverted, "state info for height %d is already reverted", fraudHeight)
+	}
+
+	//check the sequencer for this height is the same as the one in the fraud evidence
+	if stateInfo.Sequencer != seqAddr {
+		return sdkerrors.Wrapf(types.ErrWrongProposerAddr, "sequencer address %s does not match the one in the state info", seqAddr)
+	}
+
+	// slash the sequencer, clean delayed packets
+	err = k.hooks.FraudSubmitted(ctx, rollappID, fraudHeight, seqAddr)
+	if err != nil {
+		return err
+	}
+
+	//mark the rollapp as frozen. revert all pending states to finalized
+	rollapp.Frozen = true
+	k.SetRollapp(ctx, rollapp)
+
+	lastFinalizedIdx, _ := k.GetLatestFinalizedStateIndex(ctx, rollappID)
+	finalizedStateInfo, _ := k.GetStateInfo(ctx, rollappID, lastFinalizedIdx.Index)
+
+	lastFinalizedHeight := finalizedStateInfo.StartHeight + finalizedStateInfo.NumBlocks
+
+	//iterate over all height from the last finalized height to the current height
```

Comment:
possibly cleaner to move the state revert  logic to a separate function (maybe on the rollapp struct)